### PR TITLE
Mypy lifespan

### PR DESCRIPTION
**Pull request by euri10**
In the process of typing uvicorn, this does the lifespan module

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -5,10 +5,12 @@ max-line-length = 88
 [mypy]
 disallow_untyped_defs = True
 ignore_missing_imports = True
+follow_imports = silent
```
Comment (florimondmanca) [inline setup.cfg:8]: Doesn't mypy allow a `files` option here? (Maybe it doesn't, but I think it does.) It would then be picked up by IDEs too. :)

File: uvicorn/_types.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    Sequence,
+    Tuple,
+    Type,
+    Union,
+)
+
+if sys.version_info < (3, 8):
+    if TYPE_CHECKING:
+        from typing_extensions import Literal, TypedDict
+    else:  # due to different mypy warnings raised during CI for python 3.7 and 3.8
```
Comment (florimondmanca) [inline uvicorn/_types.py]: This looks a bit off. Are you able to make the topmost try/except pass mypy by adding `# type: ignore` for any of the imports?

File: uvicorn/_types.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+Sockets = Optional[List[socket.socket]]
+
+AutoHTTPProtocolType = Type[Union["H11Protocol", "HttpToolsProtocol"]]
+
+AutoWebSocketsProtocolType = Type[Union["WebSocketProtocol", "WSProtocol"]]
+
+StrPath = Union[str, PathLike]
+
+TransportType = Union["TCPTransport", Transport]
```
Comment (florimondmanca) [inline uvicorn/_types.py]: Should we strip down these types to only what's necessary for annotating lifespan?

Comment (euri10) [reply uvicorn/_types.py]: done

File: uvicorn/_types.py (hunk 2 above)
Comment (euri10) [reply uvicorn/_types.py]: I dont understand what you mean :)

File: setup.cfg (hunk 1 above)
Comment (euri10) [reply setup.cfg:8]: done

File: uvicorn/_types.py (hunk 2 above)
Comment (euri10) [reply uvicorn/_types.py]: well maybe like this, not sure that's what you meant !

**Comment by euri10** [comment]
ok I had it wrong on first pass, that's why reviews are so good !

this now uses typed lifespan messages in place of the http and ws protocol message before, and using those revealed 

1. that we didn't have a proper lifespan scope, it was missing the asgi spec, I corrected it
2. there is no use that is made of the "lifespan.shutdown.failed" in the `async def send`. Is that an omission ? Readin the spec we should deal with it, not sure how yet, any ideas ?

File: uvicorn/lifespan/on.py [hunk 4]
```diff
@@ -33,18 +37,21 @@ async def startup(self):
...
         self.logger.info("Application shutdown complete.")
 
-    async def main(self):
+    async def main(self) -> None:
         try:
             app = self.config.loaded_app
-            scope = {"type": "lifespan"}
+            scope: LifespanScope = {
+                "type": "lifespan",
+                "asgi": {"version": self.config.asgi_version, "spec_version": "2.0"},
```
Comment (florimondmanca) [inline uvicorn/lifespan/on.py:53]: Interesting! Should we submit this as a separate pull request, so that this focuses strictly on adding type hints?

**Comment by florimondmanca** [comment]
> there is no use that is made of the "lifespan.shutdown.failed" in the async def send. Is that an omission ? Readin the spec we should deal with it, not sure how yet, any ideas ?

True: https://asgi.readthedocs.io/en/latest/specs/lifespan.html#shutdown-failed-send-event

Probably let's track this as an issue? Interesting that this hasn't come up before - maybe application frameworks (eg Starlette) don't send this either?

File: uvicorn/lifespan/on.py (hunk 4 above)
Comment (euri10) [reply uvicorn/lifespan/on.py:53]: while I completely agree if I leave the scope as is and just add the type then mypy will fail with
`uvicorn/lifespan/on.py:51: error: Key 'asgi' missing for TypedDict "LifespanScope"`

so I PR the full dict in a separate one then once merged we rebase this one and add types ?

Comment (euri10) [reply uvicorn/lifespan/on.py:53]: ok added #754 !

**Review by florimondmanca — COMMENTED** [review]
Still need to think about this — the amount of verbosity in these types vs the benefit they bring doesn't seem clear yet…

In the meantime, do you know how come CI passes on 3.6 and 3.7, even though we haven’t added `typing_extensions` to Uvicorn’s dependencies? There’s probably a dep/dev dep that installs it for us. Should we add it explicitly?

**Comment by euri10** [comment]
> In the meantime, do you know how come CI passes on 3.6 and 3.7, even though we haven’t added `typing_extensions` to Uvicorn’s dependencies? There’s probably a dep/dev dep that installs it for us. Should we add it explicitly?

seems like pip does it ! this might be in the mypy setup for those versions ?

edit: well it's a dependency of mypy in fact, whatever the python version

```
❯ mkdir typing_extensions_36
❯ cd typing_extensions_36
❯ asdf local python 3.6.9
❯ which python
/home/lotso/.asdf/shims/python
❯ pip install mypy
Collecting mypy
  Downloading https://files.pythonhosted.org/packages/14/9d/914bb0c71292ffd7e1c9728b9366f9ea34c874f5ca12ac6005a400e4d130/mypy-0.782-cp36-cp36m-manylinux1_x86_64.whl (20.8MB)
    100% |████████████████████████████████| 20.8MB 1.8MB/s
Collecting typing-extensions>=3.7.4 (from mypy)
  Using cached https://files.pythonhosted.org/packages/0c/0e/3f026d0645d699e7320b59952146d56ad7c374e9cd72cd16e7c74e657a0f/typing_extensions-3.7.4.2-py3-none-any.whl
Collecting typed-ast<1.5.0,>=1.4.0 (from mypy)
  Using cached https://files.pythonhosted.org/packages/90/ed/5459080d95eb87a02fe860d447197be63b6e2b5e9ff73c2b0a85622994f4/typed_ast-1.4.1-cp36-cp36m-manylinux1_x86_64.whl
Collecting mypy-extensions<0.5.0,>=0.4.3 (from mypy)
  Using cached https://files.pythonhosted.org/packages/5c/eb/975c7c080f3223a5cdaff09612f3a5221e4ba534f7039db34c35d95fa6a5/mypy_extensions-0.4.3-py2.py3-none-any.whl
Installing collected packages: typing-extensions, typed-ast, mypy-extensions, mypy
Successfully installed mypy-0.782 mypy-extensions-0.4.3 typed-ast-1.4.1 typing-extensions-3.7.4.2
You are using pip version 18.1, however version 20.2.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

**Comment by florimondmanca** [comment]
> Collecting typing-extensions>=3.7.4 (from mypy)

Yup. But then it means that running Uvicorn will fail if the user doesn't have mypy installed! :-) So we ought to add it explicitly ourselves as well, I believe (with a proper `; python_version < '3.8'` environment marker).

**Comment by euri10** [comment]
> > Collecting typing-extensions>=3.7.4 (from mypy)
> 
> Yup. But then it means that running Uvicorn will fail if the user doesn't have mypy installed! :-) So we ought to add it explicitly ourselves as well, I believe (with a proper `; python_version < '3.8'` environment marker).

those eagle eyes, I didn't saw this ! :eagle:

**Comment by euri10** [comment]
> Still need to think about this — the amount of verbosity in these types vs the benefit they bring doesn't seem clear yet…

yes you're right, maybe we can merge this one and subsequent attempts on other modules to a `mypy_attempt` branch then once more has been merged get a clearer picture whether or not it's a benefit or not

File: uvicorn/_types.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+class LifeSpanReceiveShutdown(TypedDict):
+    """
+    https://asgi.readthedocs.io/en/latest/specs/lifespan.html#shutdown-receive-event
+    """
+
+    type: Literal["lifespan.shutdown"]
+
+
+LifespanReceiveMessage = Union[LifespanReceiveStartup, LifeSpanReceiveShutdown]
```
Comment (florimondmanca) [inline uvicorn/_types.py]: Could all of this be simplified like this…? (The type hints in `lifespan/off.py`/`lifespan/on.py` only use `LifespanReceiveMessage` and `LifespanSendMessage`, so we don't _actually need_ individual declarations for each individual message type.)

```python
class LifespanReceiveMessage(TypedDict):
    type: Literal["lifespan.startup", "lifespan.shutdown"]
```



(Handy trick here - `Literal` accepts multiple literal values, which mypy interprets as a `Union` of those literals.)

And then, similarly for send messages:

```python
class LifespanSendMessage(TypedDict):
    type: Literal[
        "lifespan.startup.complete",
        "lifespan.startup.failed",
        "lifespan.shutdown.complete",
    ]
```

(I know we'd need an optional `message` attribute for `startup.shutdown.failed`, but since we don't handle it anyway this is the form we need.)

File: uvicorn/_types.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Literal, TypedDict
+else:
+    from typing import Literal, TypedDict
+
+# SCOPES
+
+
+class ASGIDict(TypedDict):
```
Comment (florimondmanca) [inline uvicorn/_types.py]: Maybe this alternative naming…? (We know it's a `dict`, so we might want to indeed answer "what about ASGI does it contain?".)

```suggestion
class ASGISpecInfo(TypedDict):
```

File: uvicorn/_types.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
+import sys
+from typing import Optional, Union
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Literal, TypedDict
+else:
+    from typing import Literal, TypedDict
+
+# SCOPES
```
Comment (florimondmanca) [inline uvicorn/_types.py]: I'm not sure we need these comments - it's quite readable what each item corresponds to, imo.

File: setup.py [hunk 8]
```diff
@@ -41,11 +41,12 @@ def get_packages(package):
 )
 
 env_marker_win = "sys_platform == 'win32'"
-
+env_marker_below_38 = "python_version < '3.8'"
```
Comment (florimondmanca) [inline setup.py:44]: I find this style a bit odd, but at least it's consistent with what we already have. (I tend to rather see the markers inline, eg `"typing-extensions; python_version < '3.8'"`.)

**Review by florimondmanca — COMMENTED** [review]
Looking good! :+1: Left some ideas to simplify the types and reduce their LOC footprint.

**Comment by florimondmanca** [comment]
Wild idea: I don't know if we could eventually package these types up as an `asgi-typing` package, so that these type hints can be reused across ASGI-related projects… Since the spec is pretty stable now, it should be possible to define these once and for all. Could be an extra-thin package like [trio-typing](https://github.com/python-trio/trio-typing), for example.

File: uvicorn/_types.py (hunk 5 above)
Comment (euri10) [reply uvicorn/_types.py]: > (Handy trick here - Literal accepts multiple literal values, which mypy interprets as a Union of those literals.)

super nice, I learnt something today.

> (I know we'd need an optional message attribute for startup.shutdown.failed, but since we don't handle it anyway this is the form we need.)

I left both the message and the `lifespan.shutdown.failed` since we need message anyway for the `lifespan.startup.failed`

File: uvicorn/_types.py (hunk 6 above)
Comment (euri10) [reply uvicorn/_types.py]: done yes

File: uvicorn/_types.py (hunk 7 above)
Comment (euri10) [reply uvicorn/_types.py]: yep, my rough way to mentally keep track on where I am, shouldnt be here

File: setup.py (hunk 8 above)
Comment (euri10) [reply setup.py:44]: originally this was because we grouped windows / cython markers, I have no preference

**Review by florimondmanca — APPROVED** [review]
Lovely! :tada: I think this is ready for merging. :100:

File: uvicorn/lifespan/on.py [hunk 9]
```diff
@@ -1,24 +1,28 @@
...
+    def __init__(self, config: Config) -> None:
         if not config.loaded:
             config.load()
 
         self.config = config
         self.logger = logging.getLogger("uvicorn.error")
         self.startup_event = asyncio.Event()
         self.shutdown_event = asyncio.Event()
-        self.receive_queue = asyncio.Queue()
+        self.receive_queue: Queue = asyncio.Queue()
```
Comment (JayH5) [inline uvicorn/lifespan/on.py]: asyncio.Queue is generic, but only in the typestubs. So if you quote it you can do:
```suggestion
        self.receive_queue: "Queue[LifespanReceiveMessage]" = asyncio.Queue()
```

File: uvicorn/_types.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Literal, TypedDict
+else:
+    from typing import Literal, TypedDict
+
+
+class ASGISpecInfo(TypedDict):
+    version: str
+    spec_version: Optional[Union[Literal["2.0"], Literal["2.1"]]]
```
Comment (JayH5) [inline uvicorn/_types.py]:
```suggestion
    spec_version: Optional[Literal["2.0", "2.1"]]
```

File: uvicorn/lifespan/on.py (hunk 9 above)
Comment (JayH5) [reply uvicorn/lifespan/on.py]: Or possibly to avoid the extra/slightly redundant input:
```suggestion
        self.receive_queue: "asyncio.Queue[LifespanReceiveMessage]" = asyncio.Queue()
```

**Review by JayH5 — APPROVED** [review]
Small notes but I'm also +1 on this with/without suggestions.

File: uvicorn/lifespan/on.py (hunk 9 above)
Comment (euri10) [reply uvicorn/lifespan/on.py]: done !

File: uvicorn/_types.py (hunk 10 above)
Comment (euri10) [reply uvicorn/_types.py]: and done

**Review by florimondmanca — APPROVED** [review]
